news/scripts/news_cluster.py

```diff
@@ -0,0 +1,93 @@
+
+import pandas as pd
+from sklearn.cluster import AgglomerativeClustering
+from sklearn.feature_extraction.text import TfidfVectorizer
+
+from models.news import Cluster
+from mongo_news_db.db_connector import (delete_clusters, fetch_all,
+                                        upsert_cluster)
+
+SECONDS_IN_DAY = 24 * 3600
+
+
+def main():
+    delete_clusters()
+
+    parser = script_args()
+    args = parser.parse_args()
+    pipeline = ArticleClustering(args.threshold)
+    pipeline.run()
+
+
+def script_args():
+    parser = argparse.ArgumentParser(
+        description='Articles fetcher for StockNewsAPI', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--threshold', help='The maximum distance between articles that will be accepted in the cluster. Increasing this number decreases the size of clusters (Default=1.4)',
+                        type=float, default=1.4)
+    return parser
+
+
+class ArticleClustering:
+    def __init__(self, threshold):
+        self.threshold = threshold
+
+
+    def _extract(self):
+        """Fetches all the data from the db
+        
+        Returns:
+            pdf.DataFrame -- DataFrame representation of articles
+        """
+        return pd.DataFrame(list(fetch_all()))
+
+    def _transform(self, df):
+
+        """Clusters related articles
+        
+        Arguments:
+            df {pd.DataFrame} -- DataFrame representation of articles
+    
+        Returns:
+            pdf.DataFrame -- DataFrame representation of a clusters
+        """
+        
+        df['datetime'] = df['timestamp'].apply(
+            lambda x: datetime.datetime.fromtimestamp(int(x)))
+        df = df.sort_values(by=['datetime'])
+        vectorizer = TfidfVectorizer(
+            stop_words='english')
+        tfidf = vectorizer.fit_transform(df['text'])
+        clustering = AgglomerativeClustering(
+            n_clusters=None, distance_threshold=self.threshold).fit(tfidf.todense())
+        df['n_cluster'] = clustering.labels_
+        return df
+
+    def _load(self, df):
+        """Saves the clusters in the database
+        
+        Arguments:
+            df {pd.DataFrame} -- DataFrame representation of clusters
+        """
+        grouped_articles = [g.reset_index(drop=True) for i, g in df.groupby(
+            pd.Grouper(key='datetime', freq='{}s'.format(SECONDS_IN_DAY)))]
+        for group in grouped_articles:
+            clusters = [g.reset_index(drop=True) for i, g in group.groupby(
+                pd.Grouper(key='n_cluster'))]
+            for cluster in clusters:
+                average_timestamp = int(cluster['timestamp'].apply(int).mean())
+                del cluster['n_cluster'], cluster['datetime']
+                articles = list(cluster.T.to_dict().values())
+                clustered_articles = Cluster(
+                    articles, average_timestamp)
+                upsert_cluster(clustered_articles)
+
+    def run(self):
+        df = self._extract()
```

Comment:
should we rename this to a more relevant name, such as `doc_freq`

Reply:
df stands for dataframe but sure I can name it something more relevant

Reply:
:+1: for replacing all `df`s with `dataframe`